Sources/Implementation/DefaultNotificationCenter.swift

```diff
@@ -134,26 +161,34 @@ public class DefaultNotificationCenter: OPTNotificationCenter {
                 let event = myArgs[1] as? [String: Any] {
                 logEventListener(url, event)
             }
-        })
+        }
         
-        return incrementNotificationId()
+        return addGenericNotificationListener(notificationType: NotificationType.logEvent.rawValue, listener: listener)
     }
     
     public func removeNotificationListener(notificationId: Int) {
-        self.notificationListeners.removeValue(forKey: notificationId)
+        _listeners.performAtomic { (listeners) in
+            listeners.notificationListeners.removeValue(forKey: notificationId)
+        }
     }
     
     public func clearNotificationListeners(type: NotificationType) {
-        self.notificationListeners = self.notificationListeners.filter({$1.0 != type.rawValue})
+        _listeners.performAtomic { (listeners) in
+            listeners.notificationListeners = listeners.notificationListeners.filter({$1.0 != type.rawValue})
+        }
     }
     
     public func clearAllNotificationListeners() {
-        self.notificationListeners.removeAll()
+        _listeners.performAtomic { (listeners) in
+            listeners.notificationListeners.removeAll()
+        }
     }
     
     public func sendNotifications(type: Int, args: [Any?]) {
-        for values in notificationListeners.values where values.0 == type {
-            values.1(args)
+        _listeners.performAtomic { (listeners) in
+            for values in listeners.notificationListeners.values where values.0 == type {
```

Comment:
Holding all listener calls in the atomic for a while won't be safe. Let's grap the listeners synchronously and call them outside the atomic block.

Reply:
i agree with that if there is no problem with a removed listener possibly getting one more notify which could happen if you are iterating through a copy of the list.